apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,702 @@
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        public MetadataClientImpl(IHttpClientFactory httpClientFactory)
+        {
+            HttpClient = (httpClientFactory ?? new HttpClientFactory()).CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        private static MaintenanceStatus GetMaintenanceStatus(string result)
+        {
+            switch (result)
+            {
+                case "NONE": return MaintenanceStatus.None;
+                case "MIGRATE_ON_HOST_MAINTENANCE": return MaintenanceStatus.Migrate;
+                default: return MaintenanceStatus.Unknown;
+            }
+        }
+
+        private string GetMetadataHost()
+        {
+            var emulatorHost = Environment.GetEnvironmentVariable(EmulatorEnvironmentVariable);
+            return string.IsNullOrEmpty(emulatorHost) ? DefaultMetadataHost : emulatorHost;
+        }
+
+        private static T GetResult<T>(Func<Task<T>> operation)
+        {
+            try
+            {
+                return operation().Result;
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerExceptions.FirstOrDefault() ?? e;
+            }
+        }
+
+        private async Task<HttpResponseMessage> RequestMetadata(string relativeUrl, CancellationToken cancellationToken)
+        {
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"http://{metadataHost}/computeMetadata/v1/{relativeUrl}");
+            httpRequest.Headers.Add(MetadataFlavor, GoogleMetadataHeader);
+            return await HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static readonly Regex KeyRegex = new Regex(@"\A[a-zA-Z0-9-_]{1,127}\Z", RegexOptions.Compiled);
+        private static void ValidateCustomMetadataKey(string key)
+        {
+            if (!KeyRegex.IsMatch(key))
+            {
+                throw new ArgumentException("Invalid custom metadata key.", "key");
+            }
+        }
+
+        /// <summary>
+        /// Gets a token response which contains an access token to authorize a request synchronously.
+        /// </summary>
+        /// <exception cref="TokenResponseException">A status code other than success was returned from the server.</exception>
+        /// <returns>The token response.</returns>
+        public override TokenResponse GetAccessToken() =>
+            GetResult(() => GetAccessTokenAsync());
```

Comment:
Nit: all of these two-liners could be unwrapped to a single line. Also, you *may* be able to just use method groups, e.g.

```csharp
public override TokenResponse GetAccessToken() => GetResult(GetAccessTokenAsync);
```

The rules for type inference that govern when you can get away with this are hard to understand though, so if you need to use the lambda expression, that's fine.